src/appleseed/renderer/modeling/frame/frame.cpp

```diff
@@ -775,71 +746,49 @@ bool Frame::write_aov_images(const char* file_path) const
         const string aov_file_name = base_file_name + "." + safe_aov_name + ".exr";
         const string aov_file_path = (directory / aov_file_name).string();
 
+        // Write AOV image.
         if (!write_image(aov_file_path.c_str(), aov->get_image(), aov))
             success = false;
     }
 
     if (impl->m_save_extra_aovs)
     {
-        success = success && write_extra_aovs(
-            aov_images(),
-            impl->m_extra_aovs,
-            directory,
-            base_file_name,
-            ".exr");
+        for (size_t i = 0, e = impl->m_extra_aovs.size(); i < e; ++i)
+        {
+            const size_t image_index = impl->m_extra_aovs[i];
+            const Image& image = aov_images().get_image(image_index);
+
+            // Compute AOV image file path.
+            const string aov_name = aov_images().get_name(image_index);
+            const string safe_aov_name = make_safe_filename(aov_name);
+            const string aov_file_name = base_file_name + "." + safe_aov_name + ".exr";
+            const string aov_file_path = (directory / aov_file_name).string();
+
+            // Write AOV image.
+            if (!write_image(aov_file_path.c_str(), image))
+                success = false;
+        }
     }
 
     return success;
 }
 
 bool Frame::write_main_and_aov_images() const
 {
-    bool success = true;
-
-    // Get output filename.
     const string filepath = get_parameters().get_optional<string>("output_filename");
 
-    // Write main image.
     if (!filepath.empty())
     {
+        // Write main image.
         if (!write_main_image(filepath.c_str()))
-            success = false;
-    }
-
-    // Write AOVs.
-    for (size_t i = 0, e = aovs().size(); i < e; ++i)
-    {
-        const AOV* aov = aovs().get_by_index(i);
-
-        // Get output filename.
-        bf::path filepath = aov->get_parameters().get_optional<string>("output_filename");
-        filepath.replace_extension(".exr");
-
-        // Write AOV image.
-        if (!filepath.empty())
-        {
-            if (!write_image(filepath.string().c_str(), aov->get_image(), aov))
-                success = false;
-        }
-    }
-
-    if (impl->m_save_extra_aovs)
-    {
-        bf::path boost_file_path(filepath);
-        boost_file_path.replace_extension(".exr");
-
-        const bf::path directory = boost_file_path.parent_path();
-        const string base_file_name = boost_file_path.stem().string();
+            return false;
 
-        success = success && write_extra_aovs(
-            aov_images(),
-            impl->m_extra_aovs,
-            directory,
-            base_file_name,
-            ".exr");
+        // Write AOV images.
+        if (!write_aov_images(filepath.c_str()))
```

Comment:
Instead of using the beauty "output_filename" parameter, you need to get the filename for each AOV:
aov->get_parameters().get_optional<string>("output_filename") and save each AOV there.
We use this method in Gaffer, where AOVs can be saved anywhere with any filename.